core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
I was wondering about the eviction part. In the case where the timeout is finite I think we should be ok. The other option is to have an RPC call that informs us of this (I am not too fond of this, but we might be able to do this without too many changes by having the BMM call the ExecutorAllocationManager when the BM reports a status change)

Also I can switch the cached blocks map to a loading cache or something to keep its size finite.